src/v/serde/serde.h

```diff
@@ -0,0 +1,171 @@
+        if (
+          written_size
+          > std::numeric_limits<typename Type::envelope_size_t>::max()) {
+            throw std::system_error{
+              make_error_code(error_codes::envelope_too_big)};
+        }
+        auto const size = ss::cpu_to_le(
+          static_cast<typename Type::envelope_size_t>(written_size));
+        size_placeholder.write(
+          reinterpret_cast<char const*>(&size), sizeof(size));
+    } else if constexpr (std::is_scalar_v<Type>) {
+        auto le_t = ss::cpu_to_le(t);
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(Type));
+    } else if constexpr (reflection::is_std_vector_v<Type>) {
+        assert(t.size() <= std::numeric_limits<uint32_t>::max());
+        write(out, static_cast<uint32_t>(t.size()));
+        for (auto const& el : t) {
+            write(out, el);
+        }
+    }
+}
+
+template<typename T>
+std::decay_t<T> read(iobuf_parser& in) {
+    using Type = std::decay_t<T>;
+    auto t = Type();
+
+    if constexpr (has_serde_read<Type>) {
+        t.read(in);
+    } else if constexpr (is_envelope_v<Type>) {
+        // Read envelope header: version, compat version, and size.
+        auto const version = read<version_t>(in);
+        auto const compat_version = read<version_t>(in);
+        auto const size = read<typename Type::envelope_size_t>(in);
+        (void)version; // could be used to drop messages from old versions
+
+        if (compat_version > Type::__version) {
+            serde_log.error(
+              "read compat_version={} > {}::version={}\n ",
+              static_cast<int>(compat_version),
+              type_str<T>(),
+              static_cast<int>(Type::__version));
+            throw std::system_error{
+              make_error_code(error_codes::version_older_than_compat_version)};
+        }
+
+        if (in.bytes_left() < size) {
+            throw std::system_error{
+              make_error_code(error_codes::message_too_short)};
+        }
+
+        envelope_for_each_field(
+          t, [&](auto& f) { f = read<std::decay_t<decltype(f)>>(in); });
+    } else if constexpr (std::is_scalar_v<Type>) {
+        if (in.bytes_left() < sizeof(Type)) {
+            throw std::system_error{
+              make_error_code(error_codes::message_too_short)};
+        }
+
+        t = ss::le_to_cpu(in.consume_type<Type>());
+    } else if constexpr (reflection::is_std_vector_v<Type>) {
+        using value_type = typename Type::value_type;
+        t.resize(read<uint32_t>(in));
+        for (auto i = 0U; i < t.size(); ++i) {
+            t[i] = read<value_type>(in);
+        }
+    }
+
+    return t;
+}
+
+template<typename T>
+ss::future<std::decay_t<T>> read_async(iobuf_parser& in) {
+    using Type = std::decay_t<T>;
+    if constexpr (has_serde_async_read<Type>) {
+        auto const t = std::make_shared<Type>();
+        return t->serde_async_read(in).then([t]() { return *t; });
```

Comment:
there are a couple ways to handle this, but I don't think we need shared pointer (btw seastar has ss::lw_shared_ptr that should always be used: https://makedist.com/posts/2020/08/02/smart-pointers-in-seastar/).

```
auto t = std::make_unique<Type>();
auto p = t.get();
return p->method().finally([t=std::move(t)]{}); // maybe c++20 does left-to-right guaranteed
```

```
return ss::do_with(Type{}, [](Type& t) {
   return t.method().then([&t] {
      return std::move(t);
   });
});
```

```
Type t;
co_await t.serde_async_read(in);
co_return t;
```